File: Difficulty: Medium/Substrings with K Distinct/substrings-with-k-distinct.py
from collections import defaultdict

class Solution:
    def countSubstr (self, s, k):
        # Code here
        def count_substrings_with_at_most_k_distinct(s, k):
            freq = defaultdict(int)
            left = 0
            count = 0
        
            for right in range(len(s)):
                freq[s[right]] += 1
        
                while len(freq) > k:
                    freq[s[left]] -= 1
                    if freq[s[left]] == 0:
                        del freq[s[left]]
                    left += 1
        
                # Number of substrings ending at right
                count += right - left + 1
        
            return count
        
        if k == 0:
            return 0
        return (count_substrings_with_at_most_k_distinct(s, k) -
                count_substrings_with_at_most_k_distinct(s, k - 1))

File: Difficulty: Medium/Substrings with K Distinct/substrings-with-k-distinct.cpp
class Solution {
public:
    int countAtMostK(string& s, int k) {
        int n = s.size(), l = 0, r = 0, cnt = 0, unique = 0;
        vector<int> freq(256, 0); 
        while (r < n) {
            if (freq[s[r]] == 0) unique++;
            freq[s[r]]++;

            while (unique > k) {
                freq[s[l]]--;
                if (freq[s[l]] == 0) unique--;
                l++;
            }

            cnt += r - l + 1;
            r++;
        }

        return cnt;
    }

    int countSubstr(string& s, int k) {
        if (k == 0) return 0;
        return countAtMostK(s, k) - countAtMostK(s, k - 1);
    }
};